### PR TITLE
PowerShell Script

### LINKED ISSUE #313: PowerShell operations
Add basic PowerShell operations

### PR DESCRIPTION
**Pull request by gituser173**
Fixes #313 
Signed-off-by: Claudiu Ban claudiu-laurian.ban@hp.com

### CODE REVIEW

**Comment by orius123** [comment]
Do the file reading inside the open block. So you won't end up locking files.

File: content/io/cloudslang/powershell/powershell_file_script.sl [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+      return_result = r.std_out
+      status_code = r.status_code
+      error_message = r.std_err
+  outputs:
+    - return_result
+    - status_code
+    - error_message
+  results:
+    - SUCCESS
+    - FAILURE
```
Comment (Bonczidai) [inline content/io/cloudslang/powershell/powershell_file_script.sl]: Don't need to specify default results (both flow & description)

File: content/io/cloudslang/powershell/powershell_file_script.sl [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+#   The Apache License is available at
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+####################################################
+# Executes PowerShell script on a given host.
+#
+# Inputs:
+#   - host - the hostname or IP address of the PowerShell host
+#   - username - the username to use when connecting to the server
+#   - password - The password to use when connecting to the server
```
Comment (Bonczidai) [inline content/io/cloudslang/powershell/powershell_file_script.sl]: PKF ?

File: content/io/cloudslang/powershell/powershell_file_script.sl [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+#   - path_to_script - path to the script to execute on the PowerShell host
+# Outputs:
+#   - return_result - output of the powershell script
+#   - status_code - status code of the execution
+#   - error_message - error
```
Comment (Bonczidai) [inline content/io/cloudslang/powershell/powershell_file_script.sl]: `possible error message`?  error is a little bit short :)

File: content/io/cloudslang/powershell/powershell_file_script.sl [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+  inputs:
+    - host
+    - username
+    - password
+    - path_to_script
+  action:
+    python_script: |
+      import winrm
+      s = winrm.Session(host, auth=(username, password))
+      f = open(path_to_script)
```
Comment (Bonczidai) [inline content/io/cloudslang/powershell/powershell_file_script.sl]: T think you should use here Python's `with open ..` construct to be sure that file will be closed properly

File: content/io/cloudslang/powershell/powershell_file_script.sl [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+  name: powershell_file_script
+  inputs:
+    - host
+    - username
+    - password
+    - path_to_script
+  action:
+    python_script: |
+      import winrm
+      s = winrm.Session(host, auth=(username, password))
```
Comment (Bonczidai) [inline content/io/cloudslang/powershell/powershell_file_script.sl]: What happens if something goes wrong? shouldn't the relevant part be in a try - catch?

File: content/io/cloudslang/powershell/powershell_script.sl [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+#   - status_code - status code of the execution
+#   - error_message - error
+# Results:
+#   - SUCCESS - execution was successful
+#   - FAILURE - execution of the powershell script failed
+####################################################
+namespace: io.cloudslang.powershell
+
+operation:
+  name: powershell_script
```
Comment (Bonczidai) [inline content/io/cloudslang/powershell/powershell_script.sl]: flow names should be verbs like: execute_powershell_script / execute_powershell_script_from file? WDYT?

File: content/io/cloudslang/powershell/powershell_script.sl [hunk 7]
```diff
@@ -0,0 +1,47 @@
+#   (c) Copyright 2014 Hewlett-Packard Development Company, L.P.
+#   All rights reserved. This program and the accompanying materials
+#   are made available under the terms of the Apache License v2.0 which accompany this distribution.
+#
+#   The Apache License is available at
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+####################################################
+# Executes PowerShell script on a given host.
```
Comment (Bonczidai) [inline content/io/cloudslang/powershell/powershell_script.sl]: Not descriptive enough, same as in upper flow

File: test/io/cloudslang/powershell/test_powershell_file_script.sl [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+          - status_code
+        navigate:
+          SUCCESS: verify_text
+          FAILURE: FAILURE
+
+    - verify_text:
+        do:
+          strings.string_equals:
+            - first_string: str(error_message)
+            - second_string: str()
```
Comment (Bonczidai) [inline test/io/cloudslang/powershell/test_powershell_file_script.sl]: I would pass first the expected value, then the actual

File: test/io/cloudslang/powershell/test_powershell_file_script.sl [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+####################################################
+namespace: io.cloudslang.powershell
+
+imports:
+  strings: io.cloudslang.base.strings
+
+flow:
+  name: test_powershell_file_script
```
Comment (Bonczidai) [inline test/io/cloudslang/powershell/test_powershell_file_script.sl]: missing test file for this flow

what is the command executed here? I think you should verify the output e.g. print and verify it was printed

File: test/io/cloudslang/powershell/test_powershell_script.inputs.yaml [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+#   http://www.apache.org/licenses/LICENSE-2.0
+####################################################
+testRunPowerShellScript:
+  inputs:
+    - host: <host>
+    - username: <username>
+    - password: <password>
+  description: "Test run a powershell script"
+  testFlowPath: io.cloudslang.powershell.test_powershell_script
+  testSuites: [powershell]
```
Comment (Bonczidai) [inline test/io/cloudslang/powershell/test_powershell_script.inputs.yaml]: it its local, would be better to put `powershell-local` -- also please look into the possibilities to automate the tests

File: test/io/cloudslang/powershell/test_powershell_script.sl [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+          FAILURE: VERIFY_TEXT_FAILURE
+
+  outputs:
+    - return_result
+    - error_message
+    - status_code
+
+  results:
+    - SUCCESS
+    - FAILURE
```
Comment (Bonczidai) [inline test/io/cloudslang/powershell/test_powershell_script.sl]: add custom result

**Comment by oritstone** [comment]
Will be implemented in Java

**Pull request closed without merging**